### PR TITLE
CatalogExportApi auto-generation

### LINKED ISSUE #430: CatalogExportApi autogeneration
### Description

Autogenerated **CatalogExportApi* classes must be placed into *generated** folder.

**AC**:
* classes must be generated into /**generated*` folder and has the same structure CatalogExportApi/Api/Data/**
* logic with classes generation moved to /**Setup/Recurring.php**
* existing command **bin/magento dto:generate**`eliminated

### PR DESCRIPTION
**Pull request by bricht**
-Added a Setup/Recurring class to autogenerate the CatalogExportApi files upon se:up
 -Removed the cliCommand
 -Removed the manually added files from CatalogExportApi/Api/Data

### Related Pull Requests
https://github.com/magento-commerce/commerce-data-export-ee/pull/3

### Fixed Issues (if relevant)

1. Fixes magento/catalog-storefront#430

### Code Review Checklist (*)

- [ ] Story AC is completed
- [ ] proposed changes correspond to [Magento Technical Vision](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html)
- [ ] new or changed code is covered with web-api/integration tests (if applicable)
- [ ] no backward incompatible changes

### CODE REVIEW

**Comment by bricht** [comment]
@magento run all tests

**Comment by bricht** [comment]
@magento run all tests

**Comment by bricht** [comment]
@magento run WebAPI Tests

**Comment by bricht** [comment]
@magento run all tests

**Comment by bricht** [comment]
@magento run WebAPI Tests

**Comment by bricht** [comment]
@magento run all tests

**Comment by slavvka** [comment]
@magento run WebAPI Tests

**Comment by RuslanKostiv1** [comment]
Build fails while installation @bricht 
```
11:22:59 [testsuitegroup0] [Progress: 369 / 1331]
11:22:59 [testsuitegroup0] Module 'Magento_CatalogExport':
11:22:59 [testsuitegroup0] [Progress: 370 / 1331]
11:22:59 [testsuitegroup0] Module 'Magento_CatalogExportApi':
11:22:59 [testsuitegroup0] Running schema recurring...
11:22:59 [testsuitegroup0] In Recurring.php line 82:
11:22:59 [testsuitegroup0] 
11:22:59 [testsuitegroup0]   [RuntimeException]
11:22:59 [testsuitegroup0]   Could not generate ExportApi DTO's Error: Class 'Nette\PhpGenerator\PhpFile
11:22:59 [testsuitegroup0]   ' not found in /var/www/html/app/code/Magento/CatalogExportApi/Setup/Recurr
11:22:59 [testsuitegroup0]   ing.php:204
11:22:59 [testsuitegroup0]   Stack trace:
11:22:59 [testsuitegroup0]   #0 /var/www/html/app/code/Magento/CatalogExportApi/Setup/Recurring.php(80):
11:22:59 [testsuitegroup0]    Magento\CatalogExportApi\Setup\Recurring->generateFiles()
11:22:59 [testsuitegroup0]   #1 /var/www/html/setup/src/Magento/Setup/Model/Installer.php(1117): Magento
11:22:59 [testsuitegroup0]   \CatalogExportApi\Setup\Recurring->install()
11:22:59 [testsuitegroup0]   #2 /var/www/html/setup/src/Magento/Setup/Model/Installer.php(895): Magento\
11:22:59 [testsuitegroup0]   Setup\Model\Installer->handleDBSchemaData()
11:22:59 [testsuitegroup0]   #3 [internal function]: Magento\Setup\Model\Installer->installSchema()
11:22:59 [testsuitegroup0]   #4 /var/www/html/setup/src/Magento/Setup/Model/Installer.php(389): call_use
11:22:59 [testsuitegroup0]   r_func_array()
11:22:59 [testsuitegroup0]   #5 /var/www/html/setup/src/Magento/Setup/Console/Command/InstallCommand.php
11:22:59 [testsuitegroup0]   (231): Magento\Setup\Model\Installer->install()
11:22:59 [testsuitegroup0]   #6 /var/www/html/vendor/symfony/console/Command/Command.php(255): Magento\S
11:22:59 [testsuitegroup0]   etup\Console\Command\InstallCommand->execute()
11:22:59 [testsuitegroup0]   #7 /var/www/html/vendor/symfony/console/Application.php(1000): Symfony\Comp
11:22:59 [testsuitegroup0]   onent\Console\Command\Command->run()
11:22:59 [testsuitegroup0]   #8 /var/www/html/vendor/symfony/console/Application.php(271): Symfony\Compo
11:22:59 [testsuitegroup0]   nent\Console\Application->doRunCommand()
11:22:59 [testsuitegroup0]   #9 /var/www/html/lib/internal/Magento/Framework/Console/Cli.php(115): Symfo
11:22:59 [testsuitegroup0]   ny\Component\Console\Application->doRun()
11:22:59 [testsuitegroup0]   #10 /var/www/html/vendor/symfony/console/Application.php(147): Magento\Fram
11:22:59 [testsuitegroup0]   ework\Console\Cli->doRun()
11:22:59 [testsuitegroup0]   #11 /var/www/html/bin/magento(23): Symfony\Component\Console\Application->r
11:22:59 [testsuitegroup0]   un()
11:22:59 [testsuitegroup0]   #12 {main}
```

**Review by RuslanKostiv1 — CHANGES_REQUESTED** [review]
please take a look into build logs

**Comment by mslabko** [comment]
@magento run WebAPI Tests

**Comment by mslabko** [comment]
@magento run WebAPI Tests

**Comment by bricht** [comment]
@magento run all tests

**Comment by bricht** [comment]
@magento run WebAPI Tests

File: app/code/Magento/CatalogExportApi/Setup/Recurring.php [hunk 1]
```diff
@@ -62,85 +38,66 @@ class GenerateDTOFiles extends Command
...
-     * @param InputInterface $input
-     * @param OutputInterface $output
-     * @throws LocalizedException
-     * @throws RuntimeException
+     * @throws \RuntimeException
      * @throws FileSystemException
-     * @return int|void
      */
-    protected function execute(InputInterface $input, OutputInterface $output): ?int
+    public function install(SchemaSetupInterface $setup, ModuleContextInterface $context): void
```
Comment (mslabko) [inline app/code/Magento/CatalogExportApi/Setup/Recurring.php]: Please, create a separate class with "generation" logic - so we can use it inside Recurring/di compile ...

**Comment by bricht** [comment]
setup:di:compile does not work with this currently, because it does not regenerate our DTOS.
-No way to get around deletion of the files I think. \Magento\Setup\Console\CompilerPreparation::handleCompilerEnvironment deletes the whole generated folder.

-If we do the following then it works.
Add our generator class here
\Magento\Setup\Module\Di\App\Task\OperationFactory::$operationsDefinitions
Add our generator class here
\Magento\Setup\Console\Command\DiCompileCommand::getOperationsConfiguration

Obviously move the generator logic out of recurring and into another class which can be reused.

**Comment by le0n4ik** [comment]
You can use following plugin to intercept the GenerateApplicationCode operation:
```
Index: app/code/Magento/CatalogExport/Plugin/GenerateApplicationCode.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/code/Magento/CatalogExport/Plugin/GenerateApplicationCode.php	(date 1606938749913)
+++ app/code/Magento/CatalogExport/Plugin/GenerateApplicationCode.php	(date 1606938749913)
@@ -0,0 +1,27 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\CatalogExport\Plugin;
+
+use Magento\Setup\Module\Di\App\Task\Operation\ApplicationCodeGenerator;
+
+/**
+ * @see \Magento\Setup\Module\Di\App\Task\Operation\ApplicationCodeGenerator
+ */
+class GenerateApplicationCode
+{
+    /**
+     * @param ApplicationCodeGenerator $subject
+     * @return void
+     */
+    public function afterDoOperation(
+        ApplicationCodeGenerator $subject
+    ): void {
+        // TODO: Implement plugin method.
+        $result = true;
+    }
+}
\ No newline at end of file
Index: app/code/Magento/CatalogExport/etc/di.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/code/Magento/CatalogExport/etc/di.xml	(revision 47a365a72c81af47bab4b51f1196b4a95c285e69)
+++ app/code/Magento/CatalogExport/etc/di.xml	(date 1606944914618)
@@ -52,4 +52,7 @@
             </argument>
         </arguments>
     </type>
+    <type name="Magento\Setup\Module\Di\App\Task\Operation\ApplicationCodeGenerator">
+        <plugin name="GenerateGrpcCode" type="Magento\CatalogExport\Plugin\GenerateApplicationCode" sortOrder="0"/>
+    </type>
 </config>

```

**Comment by bricht** [comment]
@magento run all tests

**Comment by bricht** [comment]
@magento run all tests

**Comment by bricht** [comment]
@magento run all tests

**Comment by bricht** [comment]
@magento run all tests

**Comment by bricht** [comment]
@magento run all tests

**Comment by bricht** [comment]
The WebAPI test failure has been fixed in https://github.com/magento/catalog-storefront/pull/436 . Test is failing because the order of variants is not consistent, so it was using wrong variant to match.

The Static test failure doesnt make a whole lot of sense. It is saying that the generated getter/setter return type is invalid, which I expect is because the returned class might not exist at the time the test is run on the class, however it does exist by the end of generation. 

For example:

```
Return typehint of method                                              
Magento\CatalogExportApi\Api\Data\Product::getImage() has invalid      
type Magento\CatalogExportApi\Api\Data\ImageDeprecated. 
```

**Comment by bricht** [comment]
The rest of the failures are unrelated to my changes from what I can tell

**Comment by bricht** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run Static Tests

**Comment by RuslanKostiv1** [comment]
@magento run Static Tests

**Comment by RuslanKostiv1** [comment]
@magento run Static Tests, Magento Health Index

**Comment by RuslanKostiv1** [comment]
@magento run Static Tests, Magento Health Index

**Comment by RuslanKostiv1** [comment]
@magento import pr to magento-commerce/commerce-data-export

**Comment by magento-engcom-team** [comment]
@RuslanKostiv1 the pull request successfully imported.

**Pull request merged**